libnopegl/src/node_animkeyframe.c

```diff
@@ -319,6 +323,175 @@ static easing_type back_derivative(easing_type t, easing_type s)
+    return fabsf(s - r) > FLT_EPSILON ? NAN : s;
+}
+
+/*
+ * Fast cbrt adapted from the Android Compose library.
+ *
+ * See: https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:compose/ui/ui-util/src/commonMain/kotlin/androidx/compose/ui/util/MathHelpers.kt;l=160
+ */
+static float fast_cbrt(float x)
+{
+    union { float f32; uint32_t u32; } v = { .f32 = x };
+    v.u32 = v.u32 & 0x1ffffffffL;
+    v.u32 = 0x2a510554 + (v.u32 / 3);
+
+    float estimate = v.f32;
+    estimate -= (estimate - x / (estimate * estimate)) * (1.0f / 3.0f);
+    estimate -= (estimate - x / (estimate * estimate)) * (1.0f / 3.0f);
+    return estimate;
+}
+
+/*
+ * Find the first cubic root using Cardano's algorithm.
+ *
+ * See: https://pomax.github.io/bezierinfo/#yforx
+ */
+static float find_first_cubic_root(float p0, float p1, float p2, float p3)
+{
+    double a = 3.0 * (p0 - 2.0 * p1 + p2);
+    double b = 3.0 * (p1 - p0);
+    double c = p0;
+    double d = -p0 + 3.0 * (p1 - p2) + p3;
+
+    if (double_close_to(d, 0.0, FLT_EPSILON)) {
+        if (double_close_to(a, 0.0, FLT_EPSILON)) {
+            if (double_close_to(b, 0.0, FLT_EPSILON)) {
+                return NAN;
+            }
+            return clamp_root((float) (-c / b));
+        } else {
+            double q = sqrt(b * b - 4.0 * a * c);
+            double a2 = 2.0 * a;
+
+            float root = clamp_root(((float) ((q - b) / a2)));
+            if (!isnan(root))
+                return root;
+
+            return clamp_root(((float) ((-b - q) / a2)));
+        }
+    }
+
+    a /= d;
+    b /= d;
+    c /= d;
+
+    double o3 = (3.0 * b - a * a) / 9.0;
+    double q2 = (2.0 * a * a * a - 9.0 * a * b + 27.0 * c) / 54.0;
+    double discriminant = q2 * q2 + o3 * o3 * o3;
+    double a3 = a / 3.0;
+
+    // Three possible real roots
+    if (discriminant < 0.0) {
+        double mp33 = -(o3 * o3 * o3);
+        double r = sqrt(mp33);
+        double t = -q2 / r;
+        double cos_phi = NGLI_CLAMP(t, -1.0, 1.0);
+        double phi = acos(cos_phi);
+        double t1 = 2.0 * fast_cbrt((float)r);
+
+        float root = clamp_root((float)(t1 * cos(phi / 3.0) - a3));
+        if (!isnan(root))
+            return root;
+
+        root = clamp_root((float)(t1 * cos((phi + TAU_F64) / 3.0) - a3));
+        if (!isnan(root))
+            return root;
+
+        return clamp_root((float)(t1 * cos((phi + 2.0 * TAU_F64) / 3.0) - a3));
+    }
+
+    // Three real roots, but two of them are equal
+    if (discriminant == 0.0) {
+        float u1 = -fast_cbrt((float)q2);
+
+        float root = clamp_root((float)(2.0 * u1 - a3));
+        if (!isnan(root))
+            return root;
+
+        return clamp_root((float)(-u1 - a3));
+    }
+
+    // One real root, two complex roots
+    double sd = sqrt(discriminant);
+    double u1 = fast_cbrt((float)(-q2 + sd));
+    double v1 = fast_cbrt((float)(q2 + sd));
+
+    return clamp_root((float)(u1 - v1 - a3));
```

Comment:
Note that in my experience the analytic cubic is filled with numerical instability and I wouldn't recommend it... I ended up dropping it from the distmap for a more stable iterative algorithm